File: src/de/felixbruns/jotify/cache/Cache.java
package de.felixbruns.jotify.cache;

import java.util.List;

/**
 * Interface for cache implementations.
 * 
 * @author Felix Bruns <felixbruns@web.de>
 */
public interface Cache {
	/**
	 * Clear the entire cache.
	 */
	public void clear();
	
	/**
	 * Clear the cache for the specified category.
	 * 
	 * @param category A cache category to clear.
	 */
	public void clear(String category);
	
	/**
	 * Check if the cache contains an item. 
	 * 
	 * @param category The cache category to check.
	 * @param hash     The hash of the item to check.
	 * 
	 * @return {@code true} if it contains that item, {@code false} otherwise.
	 */
	public boolean contains(String category, String hash);
	
	/**
	 * Load data from the cache.
	 * 
	 * @param category The cache category to load from.
	 * @param hash     The hash of the item to load.
	 * 
	 * @return Cached data or {@code null}.
	 */
	public byte[] load(String category, String hash);
	
	/**
	 * Remove a single item from the cache.
	 * 
	 * @param category The cache category to remove from.
	 * @param hash     The hash of the item to remove.
	 */
	public void remove(String category, String hash);
	
	/**
	 * Store data in the cache.
	 * 
	 * @param category The cache category to store to.
	 * @param hash     The hash of the item to store.
	 * @param data     The data to store.
	 */
	public void store(String category, String hash, byte[] data);
	
	/**
	 * Store data in the cache.
	 * 
	 * @param category The cache category to store to.
	 * @param hash     The hash of the item to store.
	 * @param data     The data to store.
	 * @param size     The size of the data.
	 */
	public void store(String category, String hash, byte[] data, int size);
	
	/**
	 * List data in a cache category.
	 * 
	 * @param category The cache category to list.
	 * 
	 * @return A {@link List} of cache hashes.
	 */
	public List<String> list(String category);
}

File: src/de/felixbruns/jotify/gateway/resources/cache.js
/*
 *
 * Copyright (c) 2009, Felix Bruns <felixbruns@web.de>
 *
 */

function Cache(name){
	/* Set default values for parameters. */
	this.name = name || 'default';
	
	/* Create cache if it doesn't exist. */
	if(typeof(localStorage[this.name]) == 'undefined' || localStorage[this.name] == null){
		localStorage[this.name] = '{}';
	}
}

/* Clear. */
Cache.prototype.clear = function(){
	localStorage[this.name] = '{}';
};

/* Contains. */
Cache.prototype.contains = function(hash){
	var cache = JSON.parse(localStorage[this.name]);
	
	return typeof(cache[hash]) != 'undefined';
};

/* Load. */
Cache.prototype.load = function(hash){
	var cache = JSON.parse(localStorage[this.name]);
	
	return cache[hash];
};

/* Load all. */
Cache.prototype.loadAll = function(){
	return JSON.parse(localStorage[this.name]);
};

/* Remove. */
Cache.prototype.remove = function(hash){
	var cache = JSON.parse(localStorage[this.name]);
	
	delete cache[hash];
	
	localStorage[this.name] = JSON.stringify(cache);
};

/* Store. */
Cache.prototype.store = function(hash, data){
	var cache = JSON.parse(localStorage[this.name]);
	
	cache[hash] = data;
	
	localStorage[this.name] = JSON.stringify(cache);
};
